File: day_3/e.js
const fs = require("fs");

const exercise_1 = (inputs) => {
    let gammaRate = "";
    let epsilonRate = "";

    for (let i = 0; i < inputs[0].length; i++) {
        let bit1Counter = inputs.reduce((acc, row) => acc + Number(row[i]), 0);

        if (bit1Counter > inputs.length / 2) {
            gammaRate += "1";
            epsilonRate += "0";
        } else {
            gammaRate += "0";
            epsilonRate += "1";
        }
    }

    return parseInt(gammaRate, 2) * parseInt(epsilonRate, 2)
}

const exercise_2 = (inputs) => {
    let oxygenGeneratorRatingValues = [...inputs];
    let CO2ScrubberRatingValues = [...inputs];

    for (let i = 0; i < inputs[0].length; i++) {
        if (oxygenGeneratorRatingValues.length > 1) {
            let bit1Counter = oxygenGeneratorRatingValues.reduce((acc, row) => acc + Number(row[i]), 0);
            if (bit1Counter < oxygenGeneratorRatingValues.length / 2) {
                oxygenGeneratorRatingValues = oxygenGeneratorRatingValues.filter(v => v[i] === "1");
            } else {
                oxygenGeneratorRatingValues = oxygenGeneratorRatingValues.filter(v => v[i] === "0");
            }
        }
        if (CO2ScrubberRatingValues.length > 1) {
            let bit1Counter = CO2ScrubberRatingValues.reduce((acc, row) => acc + Number(row[i]), 0);
            if (bit1Counter < CO2ScrubberRatingValues.length / 2) {
                CO2ScrubberRatingValues = CO2ScrubberRatingValues.filter(v => v[i] === "0");
            } else {
                CO2ScrubberRatingValues = CO2ScrubberRatingValues.filter(v => v[i] === "1");
            }
        }
    }
    return parseInt(oxygenGeneratorRatingValues[0], 2) * parseInt(CO2ScrubberRatingValues[0], 2);
}

const input = fs.readFileSync("input.txt", { encoding: "utf-8" });

const inputs = input.trim().split("\n");

console.log("Exercise 1:", exercise_1(inputs));
console.log("Exercise 2:", exercise_2(inputs));

File: day_3/e.go
package main

import (
	"os"
	"strconv"
	"strings"
)

func countBit1OccurrencesAtIndex(rows []string, index int) int {
	if len(rows) == 0 {
		return 0
	}
	return (int(rows[0][index]) - '0') + countBit1OccurrencesAtIndex(rows[1:], index)
}

func exercise_1(inputs []string) int64 {
	gammaRate, epsilonRate := "", ""

	for idx := range inputs[0] {
		bit1Counter := countBit1OccurrencesAtIndex(inputs, idx)

		if bit1Counter > len(inputs)/2 {
			gammaRate += "1"
			epsilonRate += "0"
		} else {
			gammaRate += "0"
			epsilonRate += "1"
		}
	}

	gammaRateD, err := strconv.ParseInt(gammaRate, 2, 0)
	if err != nil {
		panic(err)
	}
	epsilonRateD, err := strconv.ParseInt(epsilonRate, 2, 0)
	if err != nil {
		panic(err)
	}
	return gammaRateD * epsilonRateD
}

func exercise_2(inputs []string) int64 {
	oxygenGeneratorRatingValues, CO2ScrubberRatingValues := inputs, inputs

	filter := func(slice []string, index int, char string) []string {
		newSlice := make([]string, 0)
		for _, row := range slice {
			if string(row[index]) == char {
				newSlice = append(newSlice, row)
			}
		}
		return newSlice
	}

	for idx := range inputs[0] {
		if len(oxygenGeneratorRatingValues) > 1 {
			bit1Counter := countBit1OccurrencesAtIndex(oxygenGeneratorRatingValues, idx)
			if float32(bit1Counter) < float32(len(oxygenGeneratorRatingValues))/2. {
				oxygenGeneratorRatingValues = filter(oxygenGeneratorRatingValues, idx, "1")
			} else {
				oxygenGeneratorRatingValues = filter(oxygenGeneratorRatingValues, idx, "0")
			}
		}

		if len(CO2ScrubberRatingValues) > 1 {
			bit1Counter := countBit1OccurrencesAtIndex(CO2ScrubberRatingValues, idx)
			if float32(bit1Counter) < float32(len(CO2ScrubberRatingValues))/2. {
				CO2ScrubberRatingValues = filter(CO2ScrubberRatingValues, idx, "0")
			} else {
				CO2ScrubberRatingValues = filter(CO2ScrubberRatingValues, idx, "1")
			}
		}
	}

	oxygenGeneratorRating, err := strconv.ParseInt(oxygenGeneratorRatingValues[0], 2, 0)
	if err != nil {
		panic(err)
	}
	CO2ScrubberRating, err := strconv.ParseInt(CO2ScrubberRatingValues[0], 2, 0)
	if err != nil {
		panic(err)
	}
	return oxygenGeneratorRating * CO2ScrubberRating
}

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	inputs := strings.Split(strings.Trim(string(data), "\n "), "\n")

	println("Exercise 1:", exercise_1(inputs))
	println("Exercise 2:", exercise_2(inputs))
}
